build: make libuv buildable with vs2017

refack opened the pull request:
Re: #1283 

This is the minimal change needed to get this to build on VS2017, it assumes the dev has run `vsvarsall.bat` (i.e. set-up the dev environment, and chose target and host arch) So to get it to build on appveyor you need to add to the YAML:
```
before_build:
  - cmd: "C:\Program Files (x86)\Microsoft Visual Studio 15.0\VC\Auxiliary\Build\vcvarsall.bat amd64"
```

I could add another PR with VS2017 detection code...

refack commented:
/cc @gibfahn @ctaggart

saghul commented:
How does this relate to #1283 ? Can you provide some more context?

refack commented:
- This change will make `libuv` buildable on windows with VS2017.
- It does not install or setup VS2017 (as #1283 [tries to do](https://github.com/ctaggart/libuv/blob/4d4ca864d0a0312c90f1a07d77a66f770801873e/scripts/Add-VisualStudioPackages.ps1)).
- I did not change `appveyor.yml`. I put the needed change in the description. I could add it if you want.
- This does not include VS2017 detection logic (I usually use [`msvs-com-helper`](https://github.com/node4good/msvs-com-helper))
- #1283 has a [script that manipulates](https://github.com/ctaggart/libuv/blob/4d4ca864d0a0312c90f1a07d77a66f770801873e/scripts/Update-ProjectFilesToVS2017.ps1) the `GYP` output. This PR makes that redundant.

saghul commented on appveyor.yml:
```diff
@@ -1,10 +1,13 @@
 version: v1.11.0.build{build}
 
+image:
```
The changes here look like will make appveyor only use VS 2017, is that so?

refack replied:
Yes. I can add it to the matrix.

saghul replied:
Well, your PR should *add*, not replace.

refack replied:
I agree on principle... was looking for fast feedback from AppVeyor (and from you).

refack replied:
Fixed

saghul commented on appveyor.yml:
```diff
@@ -1,10 +1,13 @@
 
 matrix:
-  fast_finish: true
+  fast_finish: false
```
what is this supposed to do?

refack replied:
`true` will stop the build on first failure (making the `allow_failures:` lines contradictory)

refack commented:
Green on Appveyor [![Build status](https://ci.appveyor.com/api/projects/status/ly0utj5x1qdqs5hb/branch/vs2017?svg=true)](https://ci.appveyor.com/project/refack/libuv/branch/vs2017)

saghul commented on gyp_uv.py:
```diff
@@ -34,37 +44,47 @@ def host_arch():
-  if rc != 0:
-    print('Error running GYP')
-    sys.exit(rc)
+def config_vs2017(config_fn):
```
can we ship this file instead of writing it every time?

refack replied:
SDK version depends on dev ENV. (toolset might change in the future as well)

saghul commented on gyp_uv.py:
```diff
@@ -34,37 +44,47 @@ def host_arch():
-    common_fn  = os.path.join(uv_root, 'common.gypi')
-    options_fn = os.path.join(uv_root, 'options.gypi')
-    # we force vs 2010 over 2008 which would otherwise be the default for gyp
-    if not os.environ.get('GYP_MSVS_VERSION'):
```
why is this removed?

refack replied:
I parameterized the `base_path` and merged the two cases.

saghul replied:
I mean setting the default GYP_MSVS_VERSION

refack replied:
Ohh, yea, that does requires an explanation... I was thinking it's not the place to set a default.
On Windows `GYP_MSVS_VERSION` comes from vcbuild.bat, but indeed I didn't check the posix flow, my bad.

refack replied:
Bottom line `GYP_MSVS_VERSION` comes from detection code in `vcbuild.bat`
IMHO here it's dropped out of the blue.

saghul commented on gyp_uv.py:
```diff
@@ -34,37 +44,47 @@ def host_arch():
+
+
+  args.append(os.path.join(arg_path, 'uv.gyp'))
+  common_fn  = os.path.join(arg_path, 'common.gypi')
```
notice the difference in absolute vs relative paths, I don't recall if that was intentional or not, but it looks like it.

refack replied:
It is intentional as stated in [line 64](https://github.com/refack/libuv/blob/c7b745035369966c639456dae52b642306b1340c/gyp_uv.py#L64) but could be parameterized.

saghul commented on gyp_uv.py:
```diff
@@ -89,6 +109,13 @@ def run_gyp(args):
+    sys.exit(rc)
+
+
+if __name__ == '__main__':
```
this is unnecessary, this file is never used as a module

refack replied:
Yes, but still cleaner 😉
I'm pushing for the same change in `nodejs`

saghul replied:
Please remove it.

refack replied:
Ok.

saghul commented on gyp_uv.py:
```diff
@@ -34,37 +44,47 @@ def host_arch():
+  o = dict()
+  td = o['target_defaults'] = dict()
+  try:
+    td['msbuild_toolset'] = 'v141'
```
why do we want to do this only if the version if != than 15.0 ?

refack replied:
`os.environ.get('VisualStudioVersion') != '15.0':`  do nothing new (i.e. `return`)
I didn't see any `target_defaults` stuff, so I added it only in the VSver 15.0 case.
See my general comment.

saghul commented:
I'm at a loss here. Can you please walk me through what you are trying to accomplish here? Yes, MSVS 2017 support, but I need to know the details: why is a new file required? Why are we removing checks we used before (which I know people rely upon), etc.

refack commented:
@saghul I'll revamp this as per your comments.
Currently it has some hardcoded values such as `os.environ.get('VisualStudioVersion') == '15.0'` => `td['msbuild_toolset'] = 'v141'`
If you want I can add the COM interrogator that is a little bit more future proof and can deduce this value.

refack commented:
P.S. I have a heavy hand and I like to slash code I don't like, then let the CI keep me in check ;)

saghul commented:
> Currently it has some hardcoded values such as os.environ.get('VisualStudioVersion') == '15.0' => td['msbuild_toolset'] = 'v141'
> If you want I can add the COM interrogator that is a little bit more future proof and can deduce this value.

How is this supposed to work with MSVS 2010 or SDK 9.1, both of which we still support?

refack commented:
Fixed by your request:
* AppVeyor will do a 2x2 matrix (2015,2017)x(86, 84)
* `gyp_uv.py` will add a default to `GYP_MSVS_VERSION = 2010`. I just added some explanation to the comment
* un-refactored gyp_uv.gyp to be more module like

> How is this supposed to work with MSVS 2010 or SDK 9.1, both of which we still support?

using the same flow as before... the function will become a no-op, => config.gypi will not be generated, so it won't be added as an arg to `GYP`

refack commented:
[Green again on the 2x2](https://ci.appveyor.com/project/refack/libuv/build/v1.11.0.build17)
@saghul BTW: If you want I can set AppVeyor to check on:
* Visual Studio 2008
* Visual Studio 2010
* Visual Studio 2012
* Visual Studio 2013
* Visual Studio 2015
* Visual Studio 2017
And I think you need to choose the one version that will make a release file (and not waste time packing them on all)

refack commented:
@saghul you like the new changes?

refack commented:
@saghul ping?

saghul commented on gyp_uv.py:
```diff
@@ -34,6 +44,19 @@ def host_arch():
+    td['msbuild_toolset'] = 'v141'
+    sdk_ver = os.environ['WindowsSDKVersion'].rstrip("\\")
+    td['msvs_windows_target_platform_version'] = sdk_ver
+    write(config_fn, o)
```
please inline the write function here

refack replied:
😲

saghul commented on gyp_uv.py:
```diff
@@ -34,6 +44,19 @@ def host_arch():
+    o = dict()
+    td = o['target_defaults'] = dict()
+    td['msbuild_toolset'] = 'v141'
+    sdk_ver = os.environ['WindowsSDKVersion'].rstrip("\\")
```
use single quotes

saghul commented on gyp_uv.py:
```diff
@@ -34,6 +44,19 @@ def host_arch():
+  # we dynamically generate the file because os.environ['WindowsSDKVersion']
+  # is system dependant
+  if os.environ.get('VisualStudioVersion') == '15.0':
+    o = dict()
```
please, use a more descriptive variable name

saghul commented on gyp_uv.py:
```diff
@@ -24,6 +25,15 @@
+  filename = os.path.join(uv_root, filename)
+  print('creating %s' % filename)
+  content = pprint.pformat(data, indent=2) + '\n'
+  print (content)
```
I'd rather not print the file

refack replied:
Just copied from `nodejs`. Maybe `if verbose`?

refack commented:
@saghul changes addressed.

refack commented:
Also I could tweak appveyor.yml in the manner I described in https://github.com/libuv/libuv/pull/1284#issuecomment-291130728

refack commented:
Good news, VS2017 [landed in `GYP`]( https://chromium-review.googlesource.com/c/456360/)
PR updated, removed code to "trick" `GYP` to work.
@saghul hope you like this better.

saghul commented on vcbuild.bat:
```diff
@@ -41,6 +41,16 @@ shift
 :args-done
 
+@rem Look for a set-up Visual Studio 2017 environment
+:vs-set-2017
```
this label is unused

saghul commented on vcbuild.bat:
```diff
@@ -41,6 +41,16 @@ shift
+:vs-set-2017
+if not defined VisualStudioVersion if not "%VisualStudioVersion%"=="15.0" goto :no-vs2017-env
+:found_vs2017
+echo Found MSVS 2017
```
Please use the same wording as with other versions

saghul commented on vcbuild.bat:
```diff
@@ -41,6 +41,16 @@ shift
+@rem Look for a set-up Visual Studio 2017 environment
+:vs-set-2017
+if not defined VisualStudioVersion if not "%VisualStudioVersion%"=="15.0" goto :no-vs2017-env
+:found_vs2017
```
unused label

saghul commented on gyp_uv.py:
```diff
@@ -48,16 +48,17 @@ def run_gyp(args):
-    # we force vs 2010 over 2008 which would otherwise be the default for gyp
+    arg_path = uv_root
+    # we force vs 2010 over 2008 which would otherwise be the default for gyp,
+	# since vs2008 uses old format `.vcproj` files
```
indent looks off here

saghul commented on gyp_uv.py:
```diff
@@ -48,16 +48,17 @@ def run_gyp(args):
+
+  args.append(os.path.join(arg_path, 'uv.gyp'))
+  common_fn  = os.path.join(arg_path, 'common.gypi')
+  options_fn = os.path.join(arg_path, 'options.gypi')
```
please put these changes in a separate commit, since they are  no longer required for MSVS 2017 support.

saghul commented on appveyor.yml:
```diff
@@ -1,28 +1,51 @@
 version: v1.11.0.build{build}
 
+image:
```
I'm not sure we want this. AppVeyor is just some experimental builds we distribute. Can you please drop these changes and let's discuss them in a follow-up PR, since they are not related to getting the build going.

refack commented:
@saghul Done. PTAL.

saghul commented:
Nice and simple! I like it! Now, is this enough to build it using MSVC 2017 or is there something extra needed?

refack commented:
Make sure the script pulls the new `GYP` (delete old downloaded ver). and that the dev run `vcvarsall.bat` before.
P.S. host-arch and target-arch are deduced from ENV

saghul commented:
> and that the dev run vcvarsall.bat before.

Why can't we have this built in, as we do for other MSVC versions? (sorry if I asked before, this has changed so much over time I can't keep track)

refack commented:
> Why can't we have this built in, as we do for other MSVC versions? (sorry if I asked before, this has changed so much over time I can't keep track)

MS changed the way to find the installed instance of VS2017, they remove `VS***COMNTOOLS` env var, and now you need to call a COM server.
I have a tool that does that with only builtin windows tools https://github.com/node4good/msvs-com-helper, I'm working on [landing it in `node` ](https://github.com/nodejs/node/pull/11852/files#diff-f1d059c92a388e475daa1fbc2a88aa3aR150), but it's a little bit cumbersome.
If you want it I can integrate it (it landed in `node-gyp`, `Boost/build`, and soon in `node` ;)

refack commented:
Also @indutny took it into `gyp.js`

refack commented:
@saghul a bit of good new, MS made it simpler to detect the VS version, but it will take some time until everyone updates, for we can detect only in some of the cases (falling back to the user calling `vcvarsall.bat` by themselves)

refack commented:
@bnoordhuis I'm working on adding the detection mechanism @saghul asked for, now that MS made it simpler again to detect VS2017.

refack commented:
Re the CI: it's really a regression test, since we don't have VS2017 workers...
P.S. Any idea why is `not ok 240 - fs_chown` failing?

saghul commented:
> Any idea why is not ok 240 - fs_chown failing?

No idea, but doesn't seem related to your changes...

saghul commented:
@refack Let's add the simple way to do it, and those who haven't updated to that version would need to call vcvarsall.bat themselves. @libuv/collaborators WDYT?

refack commented:
I'll update today.

zafields commented:
@refack Thanks for your work on this! We, the folks working on the [Azure IoT Gateway SDK](https://github.com/Azure/azure-iot-gateway-sdk/), are anxiously awaiting your fix. 👍

zafields commented:
@saghul What is the status of this PR? It looks like @refack has made changes. I see you are still blocking, but I don't see any comments since the updates?

The [Azure IoT Edge SDK](https://github.com/azure/iot-edge) has a [bug](https://github.com/Azure/iot-edge/issues/218) (with work around) based on the LibUV dependency. It would be great to see this PR come in, we have a bunch of people looking forward to it.

Thanks,
Zak

cjihrig commented:
Fresh CI: https://ci.nodejs.org/view/libuv/job/libuv-test-commit/277/

@zafields have you tested the change in it's current state to verify that it fixes your problem? Unfortunately, the CI doesn't test VS2017.

zafields commented:
@cjihrig To be honest, no I have not. We have a workaround in place and have moved forward.

The immediate problem we are experiencing, is that compiling with VS2017 isn't even a condition in the `if->else` checks, so it falls to the final else which attempts to use a very old version of the `vsbuild`. I am _assuming_ (I know, I'm embarrassed even writing the word) that once LibUV can be built with VS2017, then my problem will be solved. However, I am _certain_ that my problem will not be solved until LibUV is in fact buildable with VS2017.

Mostly, I just check in on this PR from time to time, because by it's name alone it looks to be the answer to my problem. To me, it looks like @refack has done the work, and the PR has just gone cold, so I wanted to ping the reviewers.

refack commented:
> Mostly, I just check in on this PR from time to time, because by it's name alone it looks to be the answer to my problem. To me, it looks like @refack has done the work, and the PR has just gone cold, so I wanted to ping the reviewers.

Not exactly, current solution requires the dev to call `vcvars*.bat` manually. I promised to add a `vswhere` check before falling back to this. But then I got "sucked in" to the whirlwind that is being a `nodejs` Collaborator.

I'll push that now.

refack commented:
@saghul Implemented `vswhere` detection with current behaviour as fall back.If `vswhere` is not found this is the output:
```console
'vswhere' is not recognized as an internal or external command,
operable program or batch file.
Warning: Visual Studio not found
['-Dtarget_arch=ia32', '-Duv_library=static_library', '.\\uv.gyp', '-I', '.\\common.gypi', '--depth=.', '-Dhost_arch=x64']
Project files generated.
Build skipped. To build, this file needs to run from VS cmd prompt.
```

saghul commented:
CI (just in case) https://ci.nodejs.org/view/libuv/job/libuv-test-commit/280/

schlamar commented:
Just tested it locally with VS 2017 Build Tools. Everything works as expected and tests pass 👍

schlamar commented:
**Update** @saghul This will break pyuv builds! There is no way to skip building with VS 2017 if it is installed. So builds for Python < 3.5 won't work anymore as they need another VS compiler. 

Right now the correct VS compiler can be selected in pyuv by modifying the environment, the wrong `VSXXXCOMNTOOLS` variables are removed before calling vcbuild.bat ([ref](https://github.com/saghul/pyuv/blob/v1.x/setup_libuv.py#L65)).

So there should be a way to disable building with VS 2017 even if VS 2017 is installed.

schlamar commented:
Something like

    if defined LIBUV_SKIP_VSWHERE_DETECTION goto :no-vs2017-env

refack commented:
> Update @saghul This will break pyuv builds! There is no way to skip building with VS 2017 if it is installed. So builds for Python < 3.5 won't work anymore as they need another VS compiler.
> 
> Right now the correct VS compiler can be selected in pyuv by modifying the environment, the wrong VSXXXCOMNTOOLS variables are removed before calling vcbuild.bat (ref).
> 
> So there should be a way to disable building with VS 2017 even if VS 2017 is installed.

@schlamar 
I added a command line arg `skip2017`, does that help? You could also `set skip2017=1`, and it'll take.

schlamar commented:
I would prefer a more specific environment variable, `skip2017` is **very** generally. 

As command line argument it'd OK. However, I'm not sure if that's useful as you can't skip other VS versions via command line arguments.

But maybe we should change the logic and instead of skipping a user can set a preferred VS version. We could use `GYP_MSVS_VERSION` for that. If that variable is defined we check only for the selected VS version. @saghul What do you think?

schlamar commented:
Oh, my proposal with `GYP_MSVS_VERSION` won't work for pyuv as you can build Python 3.5+ extensions with VS 2015 **and** VS 2017. So you cannot select a preferred version. In this case skipping is the way to go. 

I would propose `LIBUV_SKIP_VS2017` as env variable. But maybe we should then support `LIBUV_SKIP_VSXXX` as well?

refack commented:
> I would propose LIBUV_SKIP_VS2017 as env variable.

Ack.


> But maybe we should then support LIBUV_SKIP_VSXXX as well?

That I'm leaving for a different PR, since you need to decide on a fall through policy (yes/no? maybe the var should be a positive selection like `GYP_MSVS_VERSION` not `SKIP` logic?)

refack commented:
Rebased

saghul commented:
@schlamar Thanks for catching that! As you know we remove things from the environment when building on Windows in order to pick the right compiler. Won't this work if we skip VS 2017?

saghul commented:
@refack @schlamar Let's go back a bit and re-think this. Adding a way to skip 2017 is just a hack IMHO. We should maybe read the GYP_MSVS_VERSION env variable and just try to use the requested one instead. In pyuv we'd set that.

Another approach I've been thinking about (for pyuv) is to ditch the libuv build system completely and add the libuv sources, similar to how I build pycares: https://github.com/saghul/pycares/blob/master/setup_cares.py that would simplify the build a lot, and remove Python 2 as a dependency when building on Python 3, among other things.

So, I suggest we go back to the previous scenario, fixup pyuv, and come back with a follow-up PR if we need it after the build refactor. This might be the excuse I was looking for in order to finally do it :-)

schlamar commented:
@saghul Python 3.5+ extensions can be build with VS 2017 so we should skip VS 2017 only on Python < 3.5. For Python 3.5+ VS 2017 **or** 2015 should be allowed.

The current state of the PR should work for this use case. 

I wouldn't introduce a command line arg, though :)

schlamar commented:
@saghul We cannot select a specific VS version, see my previous comment.

saghul commented:
@schlamar Wouldn't https://github.com/libuv/libuv/pull/1284#issuecomment-305099237 solve it? It would be like building any other extension, since we wouldn't be using the libuv build system.

schlamar commented:
Yes, I guess this would work and I really like your idea. 

A one liner `if defined LIBUV_SKIP_VS2017 goto ...` shouldn't hurt, though :)

refack commented:
Removed "skip" logic.
I can help with a new PR implementing "positive selection" logic

refack commented:
Removed "skip" logic.
I can help with a new PR implementing "positive selection" logic.

ganeshmahajan1985 commented:
Hi, I am trying to build but i am getting error. kindly check below details
My machine conf. win10 x64
VS 2017 enterprise
git version 2.13.0.windows.1
python 2.7.13

Please help to resolve this.

![image](https://cloud.githubusercontent.com/assets/3907941/26823204/647ef938-4adf-11e7-815e-bc1f086c1092.png)



![capture](https://cloud.githubusercontent.com/assets/3907941/26823255/92efbe2e-4adf-11e7-9c4f-e7981eb5a181.PNG)

refack commented:
Try to run the build from a "Developer Command Prompt for VS 2017"
![image](https://cloud.githubusercontent.com/assets/96947/26825318/9155bce2-4a82-11e7-8d7b-78bc862580e7.png)

schlamar commented:
@ganeshmahajan1985 It looks like you want to build v1.10.1 with VS 2017 which is not supported. You need to fetch from this PR (see https://help.github.com/articles/checking-out-pull-requests-locally/) if you want to build libuv with VS 2017.

ganeshmahajan1985 commented:
@schlamar  Interesting. 
Let me try this.
Thanks.
@refack Yes I am using developer console only. :)

zafields commented:
Hi guys,

I am just following up. It looks like all green lights and no merge. We are getting about 3 issues a week based on lack of VS2017 support with libUV.

Do you have a timeline, any status or a short description of the hang-up I can pass along to our users?

Thanks,
Zak

zafields commented:
@refack @schlamar @saghul Can you provide any insight on the status of this PR?

bzoz commented:
@schlamar @refack vs2017 support would be nice, can we make it opt-in for the time being, before pyuv gets a fix? Maybe an command line arg, like `vcbuild vc2017` ?

refack commented:
Done, PTAL.
I used as much code from `nodejs` as was reasonable so it's a tad bit noisy, but IMHO it's best
I also updated the helper

bzoz commented on tools/vswhere_usability_wrapper.cmd:
```diff
@@ -0,0 +1,33 @@
+@if not defined DEBUG_HELPER @ECHO OFF
+setlocal
+set "InstallerPath=%ProgramFiles(x86)%\Microsoft Visual Studio\Installer"
+if not exist "%InstallerPath%" set "VSWHERE=%ProgramFiles%\Microsoft Visual Studio\Installer"
```
Shouldn't this be set `"InstallerPath=..."`?

refack replied:
Thanks! I've fixed it upstream as well.

bzoz commented on vcbuild.bat:
```diff
@@ -41,10 +43,29 @@ shift
-if defined WindowsSDKDir goto select-target
-if defined VCINSTALLDIR goto select-target
+@rem Look for Visual Studio 2017
+:vs-set-2017
```
this label is unused

refack replied:
It was there just for symmetry 🤷‍♂️ . Removed, we'll bring it back for VS2019 😉

bzoz commented:
Also, since things changed quite some: @cjihrig @bnoordhuis @saghul can you reevaluate your reviews?

bzoz commented:
CI: https://ci.nodejs.org/view/libuv/job/libuv-test-commit/307/

refack commented:
Well It's building, so I assume CI fail on `win2008r2-vs2013` is unrelated?
```
not ok 282 - threadpool_cancel_getnameinfo
# exit code 3
# Output from process `threadpool_cancel_getnameinfo`:
# A
```
https://ci.nodejs.org/job/libuv-test-commit-windows/371/nodes=win2008r2-vs2013/

bzoz commented:
I guess so.

bnoordhuis commented on vcbuild.bat:
```diff
@@ -29,6 +30,7 @@ if /i "%1"=="release"      set config=Release&goto arg-ok
 if /i "%1"=="test"         set run=run-tests.exe&goto arg-ok
 if /i "%1"=="bench"        set run=run-benchmarks.exe&goto arg-ok
 if /i "%1"=="clean"        set target=Clean&goto arg-ok
+if /i "%1"=="vs2017"        set target_env=vs2017&goto arg-ok
```
Minor indent error.

refack replied:
Ack

saghul commented on vcbuild.bat:
```diff
@@ -14,10 +14,11 @@ if /i "%1"=="/?" goto help
 set config=
 set target=Build
+set target_arch=ia32
+set target_env=vs2015
```
can we call this vslegacy or something like that instead? at a glace it looks like I could set it to vs2013 if I wanted that, but that's not the case.

refack replied:
I simply removed the default value (the downstream code test works as expected with an empty or undefined value anyway)

saghul replied:
Looks good! Some docs and we're good to go here.

refack commented:
@saghul I could find any reference to VS building except for https://github.com/libuv/libuv#windows and it still holds...

saghul commented:
@refack Time to extend it then ;-)

refack commented:
> @refack Time to extend it then ;-)

How much? do you want whole blurb from `node` https://github.com/nodejs/node/blob/master/BUILDING.md#windows-1?

saghul commented:
> How much? do you want whole blurb from node https://github.com/nodejs/node/blob/master/BUILDING.md#windows-1?

Yeah, that would be great! Thanks! 🙏

refack commented:
Added, adjusted, and de-linted (based on https://github.com/nodejs/node/pull/12756)

saghul commented:
Thanks for your perseverance and patience! Landed in 19c896dd086abb351361eab54b7e88ed5ce8af08, 0953bf0dad50d0dc77fec2deefe1c5b6c5c3f020 and 80d3b95f57e4509d93b3470e60d090b0cddb216a. 🎉

cjihrig commented:
@refack is this still building successfully for you on AppVeyor? It doesn't seem like it based on https://ci.appveyor.com/project/refack/libuv/build/v1.12.0.build41. That looks like the same error I'm currently seeing while trying to cut the 1.13.0 release.

refack commented:
> @refack is this still building successfully for you on AppVeyor? It doesn't seem like it based on https://ci.appveyor.com/project/refack/libuv/build/v1.12.0.build41. That looks like the same error I'm currently seeing while trying to cut the 1.13.0 release.

Havn't looked at it in a while, looking now.
